File: app/conversation/manager.py
"""Per-call dialogue state + lead capture + live events (ported from manager.js)."""
import asyncio
import re
import time

from ..llm.assistant import reply as assistant_reply
from ..data.leads import save_lead, EMPTY_LEAD

_sessions = {}
_subscribers = set()  # set[asyncio.Queue] for the live SSE monitor


def _emit(event, snap):
    for q in list(_subscribers):
        try:
            q.put_nowait((event, snap))
        except Exception:
            pass


def subscribe() -> "asyncio.Queue":
    q = asyncio.Queue()
    _subscribers.add(q)
    return q


def unsubscribe(q):
    _subscribers.discard(q)


def _snapshot(s):
    return {
        "id": s["id"],
        "direction": s["direction"],
        "studentName": s.get("studentName") or s["lead"].get("student_name") or "",
        "phone": s.get("phone") or s["lead"].get("mobile_number") or "",
        "lead": s["lead"],
        "turns": len(s["history"]),
        "lastMessage": s["history"][-1] if s["history"] else None,
        "transcript": s["history"],
        "startedAt": s["startedAt"],
    }


def list_active_sessions():
    return [_snapshot(s) for s in _sessions.values()]


def start_session(session_id, direction="inbound", student_name="", phone=""):
    session = {
        "id": session_id,
        "direction": direction,
        "studentName": student_name,
        "phone": phone,
        "history": [],
        "lead": {**EMPTY_LEAD, "student_name": student_name, "mobile_number": phone or ""},
        "callback": None,
        "startedAt": int(time.time() * 1000),
    }
    _sessions[session_id] = session
    _emit("started", _snapshot(session))
    return session


def get_session(session_id):
    return _sessions.get(session_id)


async def greet(session_id):
    s = _sessions.get(session_id) or start_session(session_id)
    res = await assistant_reply([], {"direction": s["direction"], "studentName": s["studentName"], "lead": s["lead"]})
    s["history"].append({"role": "assistant", "content": res["reply"]})
    _emit("turn", _snapshot(s))
    return res["reply"]


async def handle_turn(session_id, user_text):
    s = _sessions.get(session_id) or start_session(session_id)
    s["history"].append({"role": "user", "content": user_text})

    res = await assistant_reply(s["history"], {
        "direction": s["direction"], "studentName": s["studentName"], "lead": s["lead"],
    })
    s["history"].append({"role": "assistant", "content": res["reply"]})
    if res.get("leadUpdates"):
        s["lead"] = {**s["lead"], **res["leadUpdates"]}

    if re.search(r"(\d{1,2})\s*(વાગ્યે|am|pm|baje|o'?clock)", user_text, re.I):
        s["callback"] = user_text

    save_lead(session_id, s["lead"], {
        "direction": s["direction"], "callback": s["callback"], "transcript": s["history"],
    })
    _emit("turn", _snapshot(s))
    return {"reply": res["reply"], "lead": s["lead"], "endCall": res.get("endCall", False)}


def end_session(session_id):
    s = _sessions.get(session_id)
    if s:
        save_lead(session_id, s["lead"], {
            "direction": s["direction"], "callback": s["callback"],
            "transcript": s["history"], "ended": True,
        })
        _emit("ended", _snapshot(s))
        _sessions.pop(session_id, None)

File: src/conversation/manager.js
import { EventEmitter } from "events";
import { reply as assistantReply } from "../llm/assistant.js";
import { saveLead, EMPTY_LEAD } from "../data/leads.js";

// ─────────────────────────────────────────────────────────────
// Holds per-call dialogue state and orchestrates a turn:
//   user text -> assistant -> reply + lead updates -> persist.
// Sessions live in memory keyed by call/stream id.
//
// `callEvents` broadcasts live activity to the admin monitor:
//   "started" | "turn" | "ended", each with a session snapshot.
// ─────────────────────────────────────────────────────────────

const sessions = new Map();
export const callEvents = new EventEmitter();
callEvents.setMaxListeners(50);

// A trimmed, serialisable view of a session for the live monitor.
function snapshot(s) {
  return {
    id: s.id,
    direction: s.direction,
    studentName: s.studentName || s.lead.student_name || "",
    phone: s.phone || s.lead.mobile_number || "",
    lead: s.lead,
    turns: s.history.length,
    lastMessage: s.history[s.history.length - 1] || null,
    transcript: s.history,
    startedAt: s.startedAt,
  };
}

export function listActiveSessions() {
  return [...sessions.values()].map(snapshot);
}

export function startSession(sessionId, { direction = "inbound", studentName = "", phone = "" } = {}) {
  const session = {
    id: sessionId,
    direction,
    studentName,
    phone,
    history: [],
    lead: { ...EMPTY_LEAD, student_name: studentName, mobile_number: phone || "" },
    callback: null,
    startedAt: Date.now(),
  };
  sessions.set(sessionId, session);
  callEvents.emit("started", snapshot(session));
  return session;
}

export function getSession(sessionId) {
  return sessions.get(sessionId);
}

// Produce the opening line without any user input (greeting).
export async function greet(sessionId) {
  const s = sessions.get(sessionId) || startSession(sessionId);
  const { reply } = await assistantReply([], { direction: s.direction, studentName: s.studentName, lead: s.lead });
  s.history.push({ role: "assistant", content: reply });
  callEvents.emit("turn", snapshot(s));
  return reply;
}

// Process one user utterance and return the assistant's spoken reply.
export async function handleTurn(sessionId, userText) {
  const s = sessions.get(sessionId) || startSession(sessionId);
  s.history.push({ role: "user", content: userText });

  const { reply, leadUpdates, endCall } = await assistantReply(s.history, {
    direction: s.direction,
    studentName: s.studentName,
    lead: s.lead,
  });

  s.history.push({ role: "assistant", content: reply });
  if (leadUpdates && Object.keys(leadUpdates).length) {
    s.lead = { ...s.lead, ...leadUpdates };
  }

  // Capture a callback time mentioned on busy outbound calls.
  const cbMatch = userText.match(/(\d{1,2})\s*(વાગ્યે|am|pm|baje|o'?clock)/i);
  if (cbMatch) s.callback = userText;

  // Persist after every turn so a dropped call still leaves a lead.
  saveLead(sessionId, s.lead, {
    direction: s.direction,
    callback: s.callback,
    transcript: s.history,
  });

  callEvents.emit("turn", snapshot(s));
  return { reply, lead: s.lead, endCall };
}

export function endSession(sessionId) {
  const s = sessions.get(sessionId);
  if (s) {
    saveLead(sessionId, s.lead, { direction: s.direction, callback: s.callback, transcript: s.history, ended: true });
    callEvents.emit("ended", snapshot(s));
    sessions.delete(sessionId);
  }
}
